apps/src/AzureTextToSpeech.js

```diff
@@ -0,0 +1,252 @@
+      request.setRequestHeader('Authorization', `Bearer ${token}`);
+      request.setRequestHeader('Content-Type', 'application/ssml+xml');
+      request.setRequestHeader(
+        'X-Microsoft-OutputFormat',
+        'audio-16khz-32kbitrate-mono-mp3'
+      );
+      request.responseType = 'arraybuffer';
+      request.onreadystatechange = () => {
+        if (request.readyState !== READY_STATE_DONE) {
+          return;
+        }
+
+        if (request.status >= 200 && request.status < 300) {
+          const soundResponse = wrappedCreateSoundResponse({
+            bytes: request.response
+          });
+          wrappedSetCachedSound(soundResponse);
+          resolve(soundResponse);
+        } else {
+          resolve(wrappedCreateSoundResponse({error: request.statusText}));
+        }
+      };
+      request.send(ssml);
+    });
+  };
+
+  /**
+   * Plays the next sound in the queue. Automatically ends playback if the SoundResponse was unsuccessful.
+   * @param {function(ArrayBuffer, Object)} play Function that plays sound bytes.
+   * @private
+   */
+  asyncPlayFromQueue_ = async play => {
+    if (this.playing) {
+      return;
+    }
+
+    const nextSoundPromise = this.dequeue_();
+    if (!nextSoundPromise) {
+      return;
+    }
+
+    this.playing = true;
+    let response = await nextSoundPromise;
+    if (response.success()) {
+      play(response.bytes.slice(0), response.playbackOptions);
+    } else {
+      response.playbackOptions.onEnded();
+    }
+  };
+
+  /**
+   * Called when a TTS sound is done playing. Set as part of this.playbackOptions_.
+   * @private
+   */
+  onSoundComplete_ = () => {
+    this.playing = false;
+    this.asyncPlayFromQueue_();
+  };
+
+  /**
+   * Generates the cache key, which is an MD5 hash of the composite key (languageCode-gender-text).
+   * We hash the composite key to avoid extra-long cache keys (as the text is part of the key).
+   * @param {string} languageCode
+   * @param {string} gender
+   * @param {string} text
+   * @returns {string} MD5 hash string
+   * @private
+   */
+  cacheKey_ = (languageCode, gender, text) => {
+    return hashString([languageCode, gender, text].join('-'));
+  };
+
+  /**
+   * Returns the cached SoundResponse if it exists.
+   * @param {string} languageCode
+   * @param {string} gender
+   * @param {string} text
+   * @returns {SoundResponse|undefined}
+   * @private
+   */
+  getCachedSound_ = (languageCode, gender, text) => {
+    const key = this.cacheKey_(languageCode, gender, text);
+    return this.cachedSounds_[key];
+  };
+
+  /**
+   * Adds the given SoundResponse to the cache.
+   * @param {string} languageCode
+   * @param {string} gender
+   * @param {string} text
+   * @param {SoundResponse} SoundResponse
+   * @private
+   */
+  setCachedSound_ = (languageCode, gender, text, soundResponse) => {
+    const key = this.cacheKey_(languageCode, gender, text);
+    this.cachedSounds_[key] = soundResponse;
```

Comment:
Is there an upper bound on the size of this cache?

Reply:
no, there isn't right now. do you know if we've implemented something similar before? i'm not sure what the best approach would be for that

Reply:
Maybe it's not an issue -- do you have an idea of how much memory these could end up consuming in a high-use session?

I can't think of an in-house example immediately.  But one idea would be to evict the least recently used entry before adding the newest.  More complex heuristics can be considered but add.. complexity.

Reply:
ah, yeah, adding a "lastAccessed" timestamp to handle something like that makes sense. i was also thinking that post-release we may want to refactor this in-memory cache to use something like elasticache instead.

i took a look in chrome's memory profile ("Memory" tab) and looked at the "retained size" of this cache to see the amount of consumed memory (outlining that just in case there's a better way to do this, since i haven't done this many times).

**10 unique 750-character (max length) strings -- 1.8MB retained size / 1% of total.**
![Screen Shot 2020-11-19 at 12 44 07 PM](https://user-images.githubusercontent.com/9812299/99722195-2c67dd80-2a65-11eb-88bb-87227978d57d.png)

**100 unique 10-character strings -- 0.8MB retained size / 1% of total.**
![Screen Shot 2020-11-19 at 12 58 14 PM](https://user-images.githubusercontent.com/9812299/99723349-ead83200-2a66-11eb-8804-7eb2de69118a.png)

Reply:
also, your question helped me uncover a bug i just introduced, so double-thank-you!

Reply:
1-2 MB doesn't seem too bad to me, if this is a typical usage pattern, so maybe we don't need to cap the cache for now.  But I'll let you make that call...

Reply:
i agree, but will cautiously watch our memory utilization after we release this feature to keep an eye on it. in CloudWatch we have a "memory utilization and process starts" graph for our frontends, which i think is the best place to monitor, but let me know if you know of other places/ways to keep an eye on this

Reply:
This is client-side memory, isn't it?  I think those graphs would be for our frontend servers.

Reply:
lol yes, that was even in my description of the graph 🤦🏻